File: frontend/src/features/chat/messageTree.ts
import type { Message, PermissionContentData } from "@/types";

/** Keep compact message grouping local to a short, uninterrupted exchange.
 *  Older messages and replies need their own author header so the timeline does
 *  not turn into an anonymous wall of text. */
export const CONSECUTIVE_MESSAGE_WINDOW_MS = 5 * 60 * 1000;
export const DISCUSSION_CONSECUTIVE_WINDOW_MS = 30 * 60 * 1000;

export function isVisuallyConsecutive(
  previous: Message,
  current: Message,
  maxGapMs = CONSECUTIVE_MESSAGE_WINDOW_MS,
): boolean {
  const previousIsNormal = !previous.msg_type || previous.msg_type === "normal";
  const currentIsNormal = !current.msg_type || current.msg_type === "normal";
  if (
    !previousIsNormal ||
    !currentIsNormal ||
    previous.sender_id !== current.sender_id ||
    previous.sender_type !== current.sender_type ||
    previous.is_deleted ||
    current.is_deleted ||
    previous.reply_to_msg_id ||
    current.reply_to_msg_id
  ) {
    return false;
  }

  const previousAt = Date.parse(previous.created_at ?? "");
  const currentAt = Date.parse(current.created_at ?? "");
  if (!Number.isFinite(previousAt) || !Number.isFinite(currentAt)) return false;

  const gap = currentAt - previousAt;
  return gap >= 0 && gap <= maxGapMs;
}

/** Discuss threads may span a slower editorial exchange. Collapse repeated
 * identity chrome only when sender and reply target are unchanged. */
export function isDiscussionConsecutive(
  previous: Message,
  current: Message,
  maxGapMs = DISCUSSION_CONSECUTIVE_WINDOW_MS,
): boolean {
  const previousIsNormal = !previous.msg_type || previous.msg_type === "normal";
  const currentIsNormal = !current.msg_type || current.msg_type === "normal";
  if (
    !previousIsNormal ||
    !currentIsNormal ||
    previous.sender_id !== current.sender_id ||
    previous.sender_type !== current.sender_type ||
    previous.reply_to_msg_id !== current.reply_to_msg_id ||
    previous.is_deleted ||
    current.is_deleted
  ) return false;

  const previousAt = Date.parse(previous.created_at ?? "");
  const currentAt = Date.parse(current.created_at ?? "");
  if (!Number.isFinite(previousAt) || !Number.isFinite(currentAt)) return false;
  const gap = currentAt - previousAt;
  return gap >= 0 && gap <= maxGapMs;
}

/** Approvals anchored to a bot turn render inside Agent steps, not as channel rows. */
export function isFoldedPermission(m: Message): boolean {
  if (m.msg_type !== "permission") return false;
  const source = (m.content_data as PermissionContentData | null | undefined)
    ?.source_msg_id;
  return typeof source === "string" && source.length > 0;
}

export function permissionSourceId(m: Message): string | null {
  const source = (m.content_data as PermissionContentData | null | undefined)
    ?.source_msg_id;
  return typeof source === "string" && source.length > 0 ? source : null;
}

export function messageSessionId(m: Message): string | null {
  const data = m.content_data as Record<string, unknown> | null | undefined;
  const sid = data?.session_id;
  return typeof sid === "string" && sid.length > 0 ? sid : null;
}

/**
 * Split a loaded message window into top-level roots and children keyed by parent.
 * A message with `reply_to_msg_id` pointing at another loaded (non-folded) message
 * is a sub-message; otherwise it stays a root (orphan reply when parent isn't loaded).
 * Folded permissions are excluded from both roots and children.
 */
export function groupMessagesByReply(messages: Message[]): {
  roots: Message[];
  childrenByParent: Map<string, Message[]>;
  byId: Map<string, Message>;
} {
  const byId = new Map<string, Message>();
  for (const msg of messages) {
    if (isFoldedPermission(msg)) continue;
    byId.set(msg.msg_id, msg);
  }

  const childrenByParent = new Map<string, Message[]>();
  const roots: Message[] = [];

  for (const msg of messages) {
    if (isFoldedPermission(msg)) continue;
    const parentId = msg.reply_to_msg_id;
    if (parentId && byId.has(parentId) && parentId !== msg.msg_id) {
      const list = childrenByParent.get(parentId);
      if (list) list.push(msg);
      else childrenByParent.set(parentId, [msg]);
    } else {
      roots.push(msg);
    }
  }

  // Keep each sibling list in channel order (stable: messages already sorted).
  for (const [, kids] of childrenByParent) {
    kids.sort((a, b) => (a.channel_seq ?? 0) - (b.channel_seq ?? 0));
  }

  return { roots, childrenByParent, byId };
}

File: apps/ios/Sources/Models/MessageTree.swift
import Foundation

/// Client-side reply nesting helpers — mirrors
/// `frontend/src/features/chat/messageTree.ts`.
enum MessageTree {
    /// Approvals anchored to a bot turn (`source_msg_id`) belong in that turn's
    /// Agent steps on web; iOS still surfaces unresolved cards as system rows,
    /// but we exclude them from the reply tree so they don't appear as roots.
    static func isFoldedPermission(_ message: MessageDto) -> Bool {
        guard message.msgType == "permission" else { return false }
        return permissionSourceId(message) != nil
    }

    static func permissionSourceId(_ message: MessageDto) -> String? {
        guard message.msgType == "permission" else { return nil }
        if let sid = message.contentData?["source_msg_id"]?.stringValue, !sid.isEmpty {
            return sid
        }
        return nil
    }

    /// Session stamped on a bot placeholder (`content_data.session_id`) for reply reuse.
    static func messageSessionId(_ message: MessageDto) -> String? {
        guard let sid = message.contentData?["session_id"]?.stringValue, !sid.isEmpty else {
            return nil
        }
        return sid
    }

    struct Grouped {
        var roots: [MessageDto]
        var childrenByParent: [String: [MessageDto]]
        var byId: [String: MessageDto]
    }

    /// Split a loaded window into top-level roots and children keyed by parent.
    /// A message with `replyToMsgId` pointing at another loaded (non-folded)
    /// message is a sub-message; otherwise it stays a root.
    static func groupByReply(_ messages: [MessageDto]) -> Grouped {
        var byId: [String: MessageDto] = [:]
        byId.reserveCapacity(messages.count)
        for message in messages where !isFoldedPermission(message) {
            byId[message.msgId] = message
        }

        var childrenByParent: [String: [MessageDto]] = [:]
        var roots: [MessageDto] = []
        roots.reserveCapacity(messages.count)

        for message in messages {
            if isFoldedPermission(message) { continue }
            if let parentId = message.replyToMsgId,
               parentId != message.msgId,
               byId[parentId] != nil
            {
                childrenByParent[parentId, default: []].append(message)
            } else {
                roots.append(message)
            }
        }

        for key in childrenByParent.keys {
            childrenByParent[key]?.sort(by: precedesInThread)
        }

        return Grouped(roots: roots, childrenByParent: childrenByParent, byId: byId)
    }

    /// Thread order. A row that is still streaming has no `channelSeq` yet and
    /// belongs last, after every row the server has already sequenced; `msgId`
    /// is the tiebreak so rows with equal (or absent) sequence numbers keep a
    /// stable position instead of reshuffling between renders.
    static func precedesInThread(_ lhs: MessageDto, _ rhs: MessageDto) -> Bool {
        let left = lhs.channelSeq ?? Int.max
        let right = rhs.channelSeq ?? Int.max
        return left == right ? lhs.msgId < rhs.msgId : left < right
    }

    /// Whether a message belongs to a discussion root (including the root itself).
    static func inDiscussionThread(_ message: MessageDto, rootId: String) -> Bool {
        message.msgId == rootId || message.threadRootMsgId == rootId
    }

    /// Overlay live WS rows (partials, permission cards) onto the REST thread.
    /// Mirrors `frontend/src/features/chat/discussionThread.ts`.
    static func mergeDiscussion(
        root: MessageDto,
        replies: [MessageDto],
        live: [MessageDto]
    ) -> [MessageDto] {
        var byId: [String: MessageDto] = [:]
        byId.reserveCapacity(replies.count + live.count + 1)
        byId[root.msgId] = root
        for reply in replies {
            byId[reply.msgId] = reply
        }

        func consider(_ message: MessageDto) {
            if let previous = byId[message.msgId] {
                if message.isPartial == true
                    || (message.channelSeq ?? 0) >= (previous.channelSeq ?? 0)
                {
                    var merged = message
                    if merged.createdAt == nil {
                        merged.createdAt = previous.createdAt
                    }
                    if merged.senderName == nil {
                        merged.senderName = previous.senderName
                    }
                    byId[message.msgId] = merged
                }
                return
            }
            if inDiscussionThread(message, rootId: root.msgId)
                || message.replyToMsgId.flatMap({ byId[$0] }) != nil
            {
                byId[message.msgId] = message
            }
        }

        for message in live {
            consider(message)
        }
        var grew = true
        while grew {
            grew = false
            for message in live {
                if byId[message.msgId] != nil { continue }
                if let parentId = message.replyToMsgId, byId[parentId] != nil {
                    byId[message.msgId] = message
                    grew = true
                }
            }
        }

        // Dictionary.values has no defined order, so this needs a total order
        // rather than one that only compares sequence numbers.
        return byId.values.sorted(by: precedesInThread)
    }
}
